chain/src/chain.rs

```diff
@@ -712,12 +712,29 @@ impl Chain {
 	) -> Result<(), Error> {
 		status.on_setup();
 
-		// Initial check based on relative heights of current head and header_head.
+		// Initial check based on body tail and current header chain.
+		// if tail is on the current header chain, the txhashset is not needed
 		{
 			let head = self.head().unwrap();
+			let tail = self.tail().unwrap_or_else(|_| head.clone());
 			let header_head = self.header_head().unwrap();
-			if header_head.height - head.height < global::cut_through_horizon() as u64 {
-				return Err(ErrorKind::InvalidTxHashSet("not needed".to_owned()).into());
+			if header_head.height - head.height < global::cut_through_horizon() as u64
+				&& tail.last_block_h != head.last_block_h
+				&& header_head.height > tail.height
+			{
+				let mut header_cursor = self.get_block_header(&header_head.prev_block_h);
+				while let Ok(header) = header_cursor {
```

Comment:
We have `is_on_current_chain()` that I think does the same thing as this (or something very similar) - 

https://github.com/mimblewimble/grin/blob/92e41fa571adf4c7a3f3064a6e4d6ffe9486fb2e/chain/src/pipe.rs#L652-L660

```
pub fn rewind_and_apply_fork(b: &Block, ext: &mut txhashset::Extension) -> Result<(), Error> {
	// extending a fork, first identify the block where forking occurred
	// keeping the hashes of blocks along the fork
	let mut fork_hashes = vec![];
	let mut current = ext.batch.get_previous_header(&b.header)?;
	while current.height > 0 && !ext.batch.is_on_current_chain(&current).is_ok() {
		fork_hashes.push(current.hash());
		current = ext.batch.get_previous_header(&current)?;
	}
```

Reply:
Here we're doing a check `whether the tail is on the current header chain`, looks like a different thing from `is_on_current_chain()`.  
The tail is 100% on the current chain, but in case a deep fork, the header chain and the local block chain diverse deeply, the body tail could be not on the current chain.
Am I right?